File: python/tabuleiro.py
class tabuleiro:

    def __init__ (self, combinacao):

        self.combinacao = list(map(int, combinacao.split()))
        self.posicaoZero = self.combinacao.index(0)
        
        tam = len(self.combinacao)

        if tam == 4:
            self.tamanho = 2
        elif tam == 9:
            self.tamanho = 3
        elif tam == 16:
            self.tamanho = 4
        else:
            raise ValueError("combinacao incompativel")
    
    def printTabuleiro (self):


        for i in range(self.tamanho):
            print("+------" * self.tamanho + "+")
            for j in range(self.tamanho):
                index = (i * self.tamanho) + j
                if self.combinacao[index] > 9:
                    print("|  ", end="")
                else:
                    print("|   ", end="")
                print ("{}  ".format(self.combinacao[index]), end="")
            print("|")
            
        print("+------" * self.tamanho + "+")
    
    def trocarPecas (self, movimento):
        aux = self.posicaoZero

        if movimento == 'd' and self.posicaoZero >= self.tamanho:
            self.posicaoZero -= self.tamanho

        elif movimento == 'u' and self.posicaoZero <= self.tamanho*(self.tamanho-1):
            self.posicaoZero += self.tamanho
        
        elif movimento == 'r' and self.posicaoZero % self.tamanho != 0:
            self.posicaoZero -= 1

        elif movimento == 'l' and (self.posicaoZero + 1) % self.tamanho != 0:
            self.posicaoZero += 1

        self.combinacao[aux], self.combinacao[self.posicaoZero] = self.combinacao[self.posicaoZero], self.combinacao[aux]

    def combinacao_correta(self):
        for i in range(self.tamanho * self.tamanho):
            if i != self.combinacao[i]:
                return False
        return True

def main():
    combinacao = input()
    jogo = tabuleiro(combinacao)
    jogo.printTabuleiro()

    movimentos = input()
    for movimento in movimentos:
        jogo.trocarPecas(movimento)
        jogo.printTabuleiro()

    print("Posição final: ", end="")
    if jogo.combinacao_correta():
        print("true")
    else:
        print("false")


if __name__ == "__main__":
    main()

File: java/Tabuleiro.java
public class Tabuleiro {
    private final int[] combinacao;
    private int posicaoZero;
    private int tamanho;

    public Tabuleiro(String combinacao) {

        String[] numeros = combinacao.split("\\s+");
        int comprimento = numeros.length;

        switch (comprimento) {
            case 4:
                this.tamanho = 2; // Tabuleiro 2x2
                break;
            case 9:
                this.tamanho = 3; // Tabuleiro 3x3
                break;
            case 16:
                this.tamanho = 4; // Tabuleiro 4x4
                break;
            default:
                throw new IllegalArgumentException("Combinação inválida");
        }

        this.combinacao = new int[comprimento];

        int i;
        for (i=0; i<comprimento; i++){
            this.combinacao[i] = Integer.parseInt(numeros[i]);
            if (this.combinacao[i] == 0){
                this.posicaoZero = i;
            }
        }
    }

    public void PrintTabuleiro(){
        int tam = (this.tamanho)*(this.tamanho);

        for (int i = 0; i < tam; i += this.tamanho) {
            System.out.println("+------".repeat(this.tamanho) + "+");
        
            for (int j = 0; j < this.tamanho; j++) {
                if(this.combinacao[i + j] > 9)
                    System.out.print("|  ");
                else
                    System.out.print("|   ");
                
                System.out.print(this.combinacao[i + j] + "  ");
            }
            System.out.println("|");
        }
        System.out.println("+------".repeat(this.tamanho) + "+");
    }

    public void TrocarPecas(char movimento){
        int i = this.posicaoZero;
        switch (movimento) {
            case 'd':
                if (this.posicaoZero < this.tamanho)
                    return;
                else
                    this.posicaoZero -= this.tamanho;
                break;

            case 'u':
                if (this.posicaoZero > (this.tamanho)*(this.tamanho - 1))
                    return;
                else
                    this.posicaoZero += this.tamanho;
                break;

            case 'r':
                if (this.posicaoZero % this.tamanho == 0 )
                    return;
                else
                    this.posicaoZero -= 1;
                break;

            case 'l':
                if ((this.posicaoZero + 1) % this.tamanho == 0 )
                    return;
                else
                    this.posicaoZero += 1;
                break;

            default:
                System.err.println("Caracter invalido");
                return;
        }

        int aux = this.combinacao[this.posicaoZero];
        this.combinacao[this.posicaoZero] = this.combinacao[i];
        this.combinacao[i] = aux;
    }

    public boolean CombinacaoCorreta(){
        for(int i=0; i<(this.tamanho)*(this.tamanho); i++){
            if (i != this.combinacao[i])
                return false;
        }
        return true;
    }
}
